compiler/src/dotty/tools/dotc/typer/ImportInfo.scala

```diff
@@ -177,28 +182,50 @@ class ImportInfo(symf: Context ?=> Symbol,
       assert(myUnimported != null)
     myUnimported
 
-  private var myUnimported: Symbol = _
+  private val isLanguageImport: Boolean = untpd.isLanguageImport(qualifier)
 
-  private var myOwner: Symbol = null
-  private var myResults: SimpleIdentityMap[TermName, java.lang.Boolean] = SimpleIdentityMap.empty
+  private var myUnimported: Symbol = _
 
-  /** Does this import clause or a preceding import clause import `owner.feature`? */
-  def featureImported(feature: TermName, owner: Symbol)(using Context): Boolean =
+  private var featureCache: SimpleIdentityMap[TermName, java.lang.Boolean] = SimpleIdentityMap.empty
 
-    def compute =
-      val isImportOwner = site.typeSymbol.eq(owner)
-      if isImportOwner && forwardMapping.contains(feature) then true
-      else if isImportOwner && excluded.contains(feature) then false
-      else
-        var c = ctx.outer
-        while c.importInfo eq ctx.importInfo do c = c.outer
-        (c.importInfo != null) && c.importInfo.featureImported(feature, owner)(using c)
-
-    if myOwner.ne(owner) || !myResults.contains(feature) then
-      myOwner = owner
-      myResults = myResults.updated(feature, compute)
-    myResults(feature)
-  end featureImported
+  /** Does this import clause or a preceding import clause enable or disable `feature`?
+   *  @param  feature   See featureImported for a description
+   *  @return Some(true)  if `feature` is imported
+   *          Some(false) if `feature` is excluded
+   *          None        if `feature` is not mentioned, or this is not a language import
+   */
+  def mentionsFeature(feature: TermName)(using Context): Option[Boolean] =
+    def test(prefix: TermName, feature: TermName): Option[Boolean] =
+      untpd.languageImport(qualifier) match
+        case Some(`prefix`) =>
+          if forwardMapping.contains(feature) then Some(true)
+          else if excluded.contains(feature) then Some(false)
+          else None
+        case _ => None
+    feature match
+      case QualifiedName(prefix, name) => test(prefix, name)
+      case _ => test(EmptyTermName, feature)
+
+  /** Does this import clause or a preceding import clause enable `feature`?
+   *
+   *  @param  feature  a possibly quailified name, e.g.
+   *                      strictEquality
+   *                      experimental.genericNumberLiterals
+   *
+   *  An excluded feature such as `strictEquality => _` in a language import
+   *  means that preceding imports are not considered and the feature is not imported.
+   */
+  def featureImported(feature: TermName)(using Context): Boolean =
+    if !featureCache.contains(feature) then
+      featureCache = featureCache.updated(feature,
+        mentionsFeature(feature) match
+          case Some(bv) => bv
+          case None =>
+            var c = ctx.outer
+            while c.importInfo eq ctx.importInfo do c = c.outer
```

Comment:
Minor: refactor the logic above into a method?

``` Scala
class Context(...):
    /** Returns an outer context that introduces an import clause
     *
     *  For a returned context `ctx`, `ctx.isImportContext == true`.
     */
    def outerImportContext: Option[Context] = ???
```

Reply:
As long as it's used only in one place I think it's better to leave it local